Application/CsrUtil/CsrUtil.c

```diff
@@ -0,0 +1,317 @@
+  }
+
+  StrnCpyS (FirstArg, ARRAY_SIZE (FirstArg), (*Argv)[1], Space - (*Argv)[1]);
+
+  NewArgs[0] = L"Self";
+  NewArgs[1] = FirstArg;
+  NewArgs[2] = ++Space;
+
+  *Argc = 3;
+  *Argv = NewArgs;
+
+  return EFI_SUCCESS;
+}
+
+#ifdef PRINT_ARGUMENTS
+STATIC
+VOID
+PrintArguments (
+  UINTN         Argc,
+  CHAR16        **Argv
+  )
+{
+  UINTN     Index;
+
+  for (Index = 0; Index < Argc; ++Index) {
+    Print (L"%u: %s\n", Index, Argv[Index]);
+  }
+}
+#endif
+
+STATIC
+VOID
+PrintUsage (
+  VOID
+  )
+{
+  Print (L"usage: csrutil <command> [<hex-value>]\n");
+  Print (L"Modify the System Integrity Protection configuration.\n");
+  Print (L"Available commands:\n");
+  Print (L"\n");
+  Print (L"    clear\n");
+  Print (L"        Clear the existing configuration.\n");
+  Print (L"    disable [<hex-value>]\n");
+  Print (L"        Disable the protection on the machine (use default 0x%x or hex value).\n", CSR_APPLE_DISABLE_SIP_DEFAULT);
+  Print (L"    enable [<hex-value>]\n");
+  Print (L"        Enable the protection on the machine (use 0 or other legal hex value).\n");
+  Print (L"    toggle [<hex-value>]\n");
+  Print (L"        Toggle the protection on the machine (use default 0x%x or hex value).\n", CSR_APPLE_DISABLE_SIP_DEFAULT);
+  Print (L"    status\n");
+  Print (L"        Display the current configuration.\n");
+  Print (L"\n");
+}
+
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS                            OldStatus;
+  EFI_STATUS                            Status;
+  UINTN                                 Data;
+  UINTN                                 Len;
+  CHAR16                                *EndPtr;
+  UINTN                                 Argc;
+  CHAR16                                **Argv;
+  CHAR16                                **OldArgv;
+  UINT32                                CsrConfig;
+  UINT32                                Attributes;
+
+  Status = GetArguments (&Argc, &Argv);
+
+  if (EFI_ERROR (Status)) {
+    Print (L"GetArguments - %r\n", Status);
+    return Status;
+  }
+
+  OldArgv = Argv;
+  Status = SplitArguments (&Argc, &Argv);
+
+#ifdef PRINT_ARGUMENTS
+  if (!EFI_ERROR (Status)) {
+    PrintArguments (Argc, Argv);
+  }
+#endif
+
+  if (EFI_ERROR (Status) || Argc < 2) {
+    PrintUsage ();
+    return EFI_SUCCESS;
+  }
+
+  //
+  // Accept valid size valid hex only
+  //
+  if (Argc > 2) {
+    Data = 0;
+    Len = StrLen (Argv[2]);
+    if (Len > sizeof (Data) * 2) {
+      Status = EFI_UNSUPPORTED;
+    } else {
+      Status = StrHexToUintnS (Argv[2], &EndPtr, &Data);
+      if (!EFI_ERROR (Status) && (EndPtr != &Argv[2][Len] || (Data & MAX_UINT32) != Data)) {
+        Status = EFI_UNSUPPORTED;
+      }
+    }
+  }
+
+  if (EFI_ERROR (Status)) {
+    PrintUsage ();
+    return Status;
+  }
+
+  if (OldArgv != Argv) {
+    //
+    // Cls if detect called as tool
+    //
+    gST->ConOut->ClearScreen (gST->ConOut);
+  }
```

Comment:
I've hacked this a bit to work as a command line tool and when called from tools - I wasn't keen on just linking it to the OC console mode lib, since then - as a command line tool - it flashes the whole screen every time you use it.

But there might be a better way than this? Or this might be okay - it's all working fine, at least on my setup.

(As it is now, when used in Tools the entry needs to be set up in TextMode to see any output but ofc works fine (with no o/p shown, and one key press required) even if not in text mode.)

Reply:
In my opinion, no tool, be that run in shell or not should clear screen. It is what the user should do. Please remove.

Reply:
Hang on - just to be clear - when this is used from Tools without that, then you get a whole stream of whatever was in the console before, then with the latest line from this output at the end. It looks pretty bad. Not at Mac to take a screenshot. Is okay? (I mean, there is no way for the user to make the screen clear, when calling this from Tools, afaik? Using TextMode does not do that.)